src/journal/FutureImpl.cc

```diff
@@ -45,8 +45,13 @@ void FutureImpl::flush(Context *on_safe) {
 
   // instruct prior futures to flush as well
   while (prev_future) {
-    Mutex::Locker locker(prev_future->m_lock);
-    prev_future = prev_future->prepare_flush(&flush_handlers);
+    FutureImplPtr future;
+    {
+      // ensure we don't destruct future while holding its lock
+      Mutex::Locker locker(prev_future->m_lock);
+      future = std::move(prev_future->prepare_flush(&flush_handlers));
+    }
+    prev_future = std::move(future);
```

Comment:
LGTM. Though wouldn't it be a little more readable if we provide both prepare_flush "locked" and "unlocked" methods and just use "unlocked" one here?